### ISSUE TITLE
API data integration into backtesting?

### ISSUE THREAD

**Issue by Prossi79** [freqtrade/freqtrade#2595]
I am working a strategy which benefit from sentiment data such as the CFGI ->> https://alternative.me/crypto/fear-and-greed-index/

The approach how to implement it in a strategy file is straightforward, but can be used for dry / live run, right?

But how could it be implemented in the FT backtest module?

**Comment by Prossi79** [comment]
PS: thats my "test" strategy working with the CFGI API

from freqtrade.strategy.interface import IStrategy
from pandas import DataFrame
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
import requests
import pandas as pd



r = requests.get('https://api.alternative.me/fng/') # returns: value value_classification  timestamp time_until_update
x = r.json()
df = pd.DataFrame(x['data']) 



class Bbperc_sentiment(IStrategy):
    """


    """

    # Minimal ROI designed for the strategy.
    # adjust based on market conditions. We would recommend to keep it low for quick turn arounds
    # This attribute will be overridden if the config file contains "minimal_roi"
    minimal_roi = {
     #  "200":  0.01,
      #  "80":  0.02,
       # "40":  0.04,
        "0":  0.04
    }

    # Optimal stoploss designed for the strategy
    stoploss = -0.1

    # startip candle count for longest period
    startup_candle_count = 100

    # Optimal ticker interval for the strategy
    ticker_interval = '1h'

    def populate_indicators(self, dataframe: DataFrame) -> DataFrame:


        # Bollinger bands
        bollinger = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe['bb_lowerband'] = bollinger['lower']
        dataframe['bb_middleband'] = bollinger['mid']
        dataframe['bb_upperband'] = bollinger['upper']

        # %B = (Current Price - Lower Band) / (Upper Band - Lower Band) 
        dataframe['bb_perc'] = (dataframe['close'] - dataframe['bb_lowerband']) / (dataframe['bb_upperband'] - dataframe['bb_lowerband'])
 
        return dataframe


    def populate_buy_trend(self, dataframe: DataFrame) -> DataFrame:
        dataframe.loc[
            (             
                    (dataframe['bb_perc'] < -0.7 ) &
                    (df['value']) < 50 # buy sentiment value
              ),
            'buy'] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame) -> DataFrame:
        dataframe.loc[
            (
                
                    (dataframe['bb_perc'] > 0 ) &
                    (df['value']) > 50 # sell sentiment value


            ),
            'sell'] = 1

        return dataframe

**Comment by hroff-1902** [comment]
formatting improved for the codeblock above

**Comment by xmatthias** [comment]
I think your implementation has a problem for dry-run too.

The place you're making the API call is out side of the strategy. This code is only run once at startup - so assuming you have the bot running for a week - then the sentiment data will never be updated.

You should place it in a function, which you call every X (probably once a day, or once an hour?)

### backtesting 

In order to use this data in backtesting, you'll need historic data - otherwise you would use today's sentiment to simulate trades a month ago - 
which will be wrong in 2 ways: 
* You're using future data which should not be available in realistic scenarios
* you're using sentiment from a completely different timeperiod (today's sentiment will also be invalid in january ...)

Is historic available from this api? (it's not from the endpoint in the code ...).
if not, it'll be difficult (you'd need to setup a job to collect this data periodically - which is not something freqtrade can help you with).

Once you have historic data (best in a dataframe, with `[date, value]` columns - you can join it to the backtesting dataframe, `ffill()` the data (assuming sentiment data has a lower frequency) - and use it as any other indicator.

**Comment by hroff-1902** [comment]
```
dataframe.loc[
            (             
                    ...
                    (df['value']) < 50 # buy sentiment value
              ),
            'buy'] = 1
```
--- I guess this is also wrong for dry-run/live, since it does not select (does not uses) the last signal value from the `df` dataframe and the datatimes in `dataframe` and `df` are not the same. `df[-1]['value']` should be used at least to work in dry-run/live and for backtesting the dataframes should be resampled to the same timeframe and aligned to contain data for the same datatimerange (as @xmatthias said)

**Comment by Prossi79** [comment]
yes, I was thinking about the "rapid prototype" and you are right .. it just takes the CFGI value if the bot starts up, thats scrap

@hroff-1902 @xmatthias  - could you assist to implement this function call into a strategy as example / template (too heavy for me ... ) ? this sentiment value is being updated just once a day.

@xmatthias - yes, historic data are available from this API. It says: "For more data (until Feb 01, 2018), please use the Fear & Greed API."  With this data backtests would make sense for sure, of course it would not make sense with just todays value :)

**Comment by hroff-1902** [comment]
I think it's better to move this issue into the freqtrade-strategies repository, since the result is assumed to be placed there.

Wouldn't you mind?

**Comment by Prossi79** [comment]
yes, no worries

Thanks and BR
